links/src/commonMain/kotlin/utils/decoders.kt

```diff
@@ -18,61 +18,192 @@
 package com.splendo.kaluga.links.utils
 
 import kotlinx.serialization.DeserializationStrategy
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.KSerializer
 import kotlinx.serialization.SerializationException
 import kotlinx.serialization.descriptors.SerialDescriptor
-import kotlinx.serialization.encoding.AbstractDecoder
 import kotlinx.serialization.encoding.CompositeDecoder
+import kotlinx.serialization.encoding.Decoder
 import kotlinx.serialization.modules.EmptySerializersModule
 import kotlinx.serialization.modules.SerializersModule
 
-internal fun <T> decodeFromList(list: List<Any>, deserializer: DeserializationStrategy<T>): T {
-    val decoder = LinksDecoder(ArrayDeque(list))
-    return decoder.decodeSerializableValue(deserializer)
-}
+internal fun <T> decodeFromMap(fields: Map<String, List<String>>, serializer: KSerializer<T>): T = MapDecoder(fields).decodeSerializableValue(serializer)
```

Comment:
@Daeda88 suggested adapting Json deserializer somehow, but I'd rather not spend even more time on the feature. all the implementation is private so it'd be easy to update it if needed